openshift/WFLY-15421_OIDC_for_WildFly_s2i_v2.adoc

```diff
@@ -0,0 +1,306 @@
+
+We are offering a new set of env variables `OIDC_*` to be used to configure the subsystem, enable the discovery of deployments and registration of 
+clients into the authorization server.
+
+The existing `SSO_*` env variables used for OIDC are usable for Keycloak OIDC integration. 
+It helps transition smoothly to use the native OIDC support. 
+This proposal defines a mapping between the `SSO_*` and `OIDC_*` env variables.
+The `SSO_*` env variables are deprecated.
+
+The env variable `OIDC_PROVIDER_NAME=<name>` MUST be set when using `OIDC_*` env variables. When using `SSO_*` env variables 
+the provider name `keycloak` is internally used.
+
+A mix of `SSO_*` and `OIDC_*` env variables, although not recommended, would work, 
+new env variables will only be added in the `OIDC_*` namespace. We have a first example `OIDC_PROVIDER_SSL_REQUIRED` 
+that would be new without a mapping in the `SSO_*` namespace.
+
+Example of configuration using new OIDC env variables for Keycloak:
+
+```
+OIDC_PROVIDER_NAME=keycloak
+OIDC_PROVIDER_URL=http://keycloak/auth/realms/Wildfly
+OIDC_USER_NAME=foo
+OIDC_USER_PASSWORD=foo
+OIDC_HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+Example of configuration using the existing SSO env variables:
+
+```
+SSO_URL=http://keycloak/auth
+SSO_REALM=Wildfly
+SSO_USER=foo
+SSO_PASSWORD=foo
+HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+Example of configuration using new OIDC env variables for a custom provider:
+
+```
+OIDC_PROVIDER_NAME=other
+OIDC_PROVIDER_URL=http://foo/specific/url/for/provider
+OIDC_HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+===== Env variables mapping table 
+
+|===
+|Env var |SSO|Details|Required|Defaults
+
+|OIDC_PROVIDER_NAME
+|NONE. When SSO_* env var are used, “keycloak” name is internally set.
+|To identify the provider to use.
+|Y
+|
+
+|OIDC_PROVIDER_URL
+|`$SSO_URL/realms/$SSO_REALM`
+|The url of the provider
+|Y
+|
+
+|OIDC_USER_NAME
+|SSO_USERNAME
+|User name used to get a token, used for dynamic client registration.
+|Y
+|
+
+|OIDC_USER_PASSWORD
+|SSO_PASSWORD
+|User password used to get a token, used for dynamic client registration.
+|Y
+|
+
+|OIDC_SECURE_DEPLOYMENT_SECRET
+|SSO_SECRET
+|Secure deployment secret known in the subsystem secure-deployment and the auth server client.
+|N
+|
+
+|OIDC_SECURE_DEPLOYMENT_PRINCIPAL_ATTRIBUTE
+|SSO_PRINCIPAL_ATTRIBUTE
+|Configure the principal value of the principal name.
+|N
+|Default to sub (token id) for keycloak, typical value is preferred_username 
+
+|OIDC_SECURE_DEPLOYMENT_ENABLE_CORS
+|SSO_ENABLE_CORS
+|Enable CORS
+|N
+|False by default 
+
+|OIDC_SECURE_DEPLOYMENT_BEARER_ONLY
+|SSO_BEARER_ONLY
+|Deployment that accepts Bearer token only, no logging.
+|N
+|False by default
+
+|OIDC_PROVIDER_SSL_REQUIRED
+|NONE, new env variable (see CLOUD-3529 )
+|Is https enabled in the realm. By default external (private and local address no https).
+|N
+|external
+
+|OIDC_PROVIDER_TRUSTSTORE
```

Comment:
Is it possible to make it somehow easier for the users, for example having just a variable referencing an OpenShift secret with the truststore (or even just trusted certificate(s))?

Reply:
I think that it is a good idea. It would avoid to mount the secret trustore in the pod. We could see it as a new RFE with the introduction of OIDC_PROVIDER_TRUSTORE_SECRET and OIDC_PROVIDER_CERTIFICATE_SECRET.

What I understand is that we would use the secret name, query the Openshift API to retrieve the Truststore secret from the current namespace then setup a local truststore file locally. I think that it would be doable. The question I am wandering are the authorizations required for a user to be able to retrieve a secret. 
I did a quick attempt using openshift sandbox and I was not able to get the secrets from the pod but this would require more investigations.

Reply:
I have not enough knowledge in this area. I just remember how painful was to configure certificates for EAP+RH-SSO on OpenShift for me a year or two ago. If such feature is worth the effort I vote for the new RFE.

I have noticed there is JGROUPS_ENCRYPT_SECRET environment variable. Maybe we can get partially inspired from the underlying solution there. See https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.4/html-single/getting_started_with_jboss_eap_for_openshift_online/index#configuring_sym_encrypt

Reply:
JGROUPS_ENCRYPT_SECRET is not used as you think it is. It is the env variable to provide the secret in the template: https://github.com/jboss-container-images/jboss-eap-openshift-templates/blob/eap74-dev/templates/eap74-https-s2i.json#L720
A user still needs to provide the name and directory: JGROUPS_ENCRYPT_KEYSTORE and JGROUPS_ENCRYPT_KEYSTORE_DIR. So actually when using helm charts, these env variables are no more in the context. We should fix that in some way.